File: newcoder/wrrite-exam/2023-08-19美团/4.cpp
#include <iostream>
#include <vector>
using namespace std;

const int mod = 1e9 + 7;

int dfs(int i, int left, vector<vector<int>>& memo, vector<int>& a, int n) {
    if (i == n) {
        if (left == 0) {
            return 1;
        }
        return 0;
    }
    if (memo[i][left] != -1) {
        return memo[i][left];
    }
    int res = 0;
    for (int j = 1; j < a[i]; j++) {
        if (left >= j) {
            res = (res + dfs(i + 1, left - j, memo, a, n)) % mod;
        }
    }
    for (int j = a[i] + 1; left - j >= n - (i + 1) && j <= left; j++) {
        res = (res + dfs(i + 1, left - j, memo, a, n)) % mod;
    }
    memo[i][left] = res;
    return res;
}

int main() {
    int n;
    cin >> n;
    vector<int> a(n);
    int s = 0;
    for (int i = 0; i < n; i++) {
        cin >> a[i];
        s += a[i];
    }

    vector<vector<int>> memo(n + 1, vector<int>(505, -1));
    int res = dfs(0, s, memo, a, n);
    cout << res << endl;

    return 0;
}

File: newcoder/wrrite-exam/2023-08-19美团/4.go
package main

import (
	"bufio"
	. "fmt"
	"io"
	"os"
)

func run(_r io.Reader, _w io.Writer) {
	in := bufio.NewReader(_r)
	out := bufio.NewWriter(_w)
	defer out.Flush()
	var n int 
	Fscan(in, &n)
	a := make([]int, n)
	s := 0
	for i := 0 ; i < n ; i ++ {
		Fscan(in, &a[i])
		s += a[i]
	}
	const mod int = 1e9 + 7 

	memo := make([][]int, n + 1)
	for i := range memo{
		memo[i] = make([]int, 505)
		for j := range memo[i]{
			memo[i][j] = -1
		}
	}
	var dfs  func(i int, left int) int
	dfs = func(i int, left int) int {
		if i == n{
			if left == 0{
				return 1
			}
			return 0
		}
		if memo[i][left] != -1{
			return memo[i][left]
		}
		res := 0
		for j := 1 ; j < a[i] ; j ++ {
			if left >= j {
				res = (res + dfs(i + 1, left - j)) % mod
			}
		}
		for j := a[i] + 1 ; left - j >= n - (i + 1) && j <= left; j ++ {
			res = (res + dfs(i + 1, left - j)) % mod
		}
		memo[i][left] = res 
		return res
	} 

	res := dfs(0, s)
	Fprintln(out, res)
	return 
}

func main() { run(os.Stdin, os.Stdout)}